streaming/base/dataset.py

```diff
@@ -765,6 +787,128 @@ def _resample_streams(self, epoch: int) -> Tuple[NDArray[np.int64], NDArray[np.i
+                # same as the ratio of the stream's samples to overall samples.
+                # This ensures that the overall training shuffle block size is still approximately
+                # equal to what is set by the user, and allows for reasoning about cache_limit as well.
+                shuffle_block_portion = int(self.shuffle_block_size * stream.proportion)
+                stream_shuffle = get_shuffle(self.shuffle_algo, shuffle_units,
+                                             self.num_canonical_nodes, self.shuffle_seed, epoch,
+                                             shuffle_block_portion)
+                stream_partition = np.where(stream_partition != -1,
+                                            stream_shuffle[stream_partition], -1)
+            # The small_per_big array already corresponds to indices of samples per shard of each stream.
+            # So each sample ID in the stream's partition already corresponds to the sample ID in the right shard.
+            partition_per_stream.append(
+                np.where(stream_partition != -1, small_per_big[stream_partition], -1))
+
+        # We now merge the partitions from each stream to get our final partition over all streams, where
+        # each global batch has samples only from a single stream.
+        # Partitions are arranged (physical nodes, ranks per node, workers per rank, batches per worker, batch size).
+        batches_per_stream = []
+        batches_from_partitions = []
+        for i, partition in enumerate(partition_per_stream):
+            # reshape the partition to be global batches in order of traversal, and count only batches without -1 in them.
+            global_batches_inorder = partition.transpose(3, 2, 0, 1, 4).reshape(
+                -1, batch_size * world.ranks_per_node * world.num_nodes)
+            num_full_batches = np.count_nonzero(np.min(global_batches_inorder, axis=1) >= 0)
+            batches_per_stream.append(num_full_batches)
+            if num_full_batches != global_batches_inorder.shape[0]:
+                logger.warning(
+                    'Because of the `per_stream` batching method, some batches with an inadequate number of samples '
+                    + 'from stream with index ' + str(i) + ' are being dropped.')
+            batches_from_partitions.append(global_batches_inorder[:num_full_batches])
+
+        # Combine all global batches from all streams into one array.
+        all_partition_batches = np.concatenate(batches_from_partitions)
+
+        # Shuffle seed changes with every epoch so that the order of streams in our batches changes as well.
+        epoch_rng = np.random.default_rng(self.shuffle_seed + epoch)
+
+        # stream_origins is an array that tells us which stream each batch is using.
+        stream_origins = np.concatenate(
+            [np.full(n_batch, i) for i, n_batch in enumerate(batches_per_stream)])
+        epoch_rng.shuffle(stream_origins)
```

Comment:
Does this shuffling logic downloads extra shard files when training runs for more than one epoch ?

Reply:
No -- we won't be downloading more shard files when training on more than one epoch. This shuffle is to determine which stream each batch will come from. So if we have 3 streams, originally the `stream_origins` array will look something like `[0,0,0,1,1,1,1,2,2]`, assuming we have 3 batches from stream 0, 4 batches from stream 1, and 2 batches from stream 2. Then we shuffle the `stream_origins` array to get the epoch batch ordering of `[1,2,1,0,1,1,0,2,0]`, for example. This ordering will change between epochs but will use the same samples/shards from each stream.